napari/components/scale_bar.py

```diff
@@ -0,0 +1,108 @@
+import numpy as np
+
+from ..utils.colormaps.standardize_color import transform_color
+from ..utils.events import EmitterGroup
+from ._viewer_constants import Position
+
+
+class ScaleBar:
+    """Scale bar indicating size in world coordinates.
+
+    Attributes
+    ----------
+    background_color : np.ndarray
+        Background color of canvas. If scale bar is not colored
+        then it has the color opposite of this color.
+    colored : bool
+        If scale bar are colored or not. If colored then
+        default color is magenta. If not colored than
+        scale bar color is the opposite of the canvas
+        background.
+    events : EmitterGroup
+        Event emitter group
+    position : str
+        Position of the scale bar in the canvas. Must be one of
+        'top left', 'top right', 'bottom right', 'bottom left'.
+        Default value is 'bottom right'.
+    ticks : bool
+        If scale bar has ticks at ends or not.
+    visible : bool
+        If scale bar is visible or not.
+    """
+
+    def __init__(self):
+
+        # Events:
+        self.events = EmitterGroup(
+            source=self,
+            auto_connect=True,
+            visible=None,
+            colored=None,
+            position=None,
+            ticks=None,
+        )
+        self._visible = False
+        self._colored = False
+        self._background_color = np.array([1, 1, 1])
+        self._ticks = True
+        self._position = Position('bottom_right')
+
+    @property
+    def visible(self):
+        """bool: If scale bar is visible or not."""
+        return self._visible
+
+    @visible.setter
+    def visible(self, visible):
+        self._visible = visible
+        self.events.visible()
+
+    @property
+    def colored(self):
+        """bool: If scale bar is colored or not."""
+        return self._colored
+
+    @colored.setter
+    def colored(self, colored):
+        self._colored = colored
+        self.events.colored()
+
+    @property
+    def background_color(self):
```

Comment:
I think that, ideally, we shouldn't store the background color of the canvas in `ScaleBar`. To avoid two sources of "truth", the background color shouldn't be duplicated here (the background color also lives in `ViewerModel.palette`)". Perhaps it would be better to connect the color update function to the ViewerModel `palette` event.

That being said, I see that we are using this pattern elsewhere (e.g., `Axes`), so I think we should proceed this way for now.

Reply:
But it is updated in palette event of viewer.

Reply:
> But it is updated in palette event of viewer.

Hey @Czaki ! It is true that the value of `ScaleBar.background_color` is updated by the `ViewerModel`. My concern here is that we are storing the value of the viewer background color in multiple places, which increases the amount of bidirectional synchronization we need to/should do. I say bidirectional synchronization because I think it would be reasonable for a user who reads the docs for `ScaleBar.background_color` (a public property) to expect to be able to change the viewer canvas background color via `ScaleBar.background_color` (thus the need for bidirectional sync). For just a couple of properties, this isn't such a big deal, but as napari grows in complexity, managing all of these events and updates becomes burdensome.

A possible alternative would be to use a function that takes the new canvas background color as an input and sets the scale bar color accordingly without storing the new background color.

All of that being said, as I mentioned above, I think we should move forward as is, since it both works and we have used that pattern elsewhere (e.g., `Axes`). I just wanted to make a note here in case we want to revisit in the future.

Reply:
Yeah, agreed @kevinyamauchi I didn't love that, something we can think about fixing for sure